pkg/core/mpt/trie.go

```diff
@@ -0,0 +1,351 @@
+package mpt
+
+import (
+	"bytes"
+	"errors"
+
+	"github.com/nspcc-dev/neo-go/pkg/core/storage"
+	"github.com/nspcc-dev/neo-go/pkg/crypto/hash"
+	"github.com/nspcc-dev/neo-go/pkg/io"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+)
+
+// Trie is an MPT trie storing all key-value pairs.
+type Trie struct {
+	storage.Store
+
+	root Node
+}
+
+// ErrNotFound is returned when requested trie item is missing.
+var ErrNotFound = errors.New("item not found")
+
+// NewTrie returns new MPT trie.
+func NewTrie(root Node, store storage.Store) *Trie {
+	if root == nil {
+		root = new(HashNode)
+	}
+
+	return &Trie{
+		Store: store,
+		root:  root,
+	}
+}
+
+// Get returns value for the provided key in t.
+func (t *Trie) Get(key []byte) ([]byte, error) {
+	path := toNibbles(key)
+	r, bs, err := t.getWithPath(t.root, path)
+	if err != nil {
+		return nil, err
+	}
+	t.root = r
```

Comment:
Why does it change the root?

Reply:
Because if `root` was a hashed node, we replace it with the new node retrieved.

Reply:
It's a bit unusual to see `Get` changing anything in the structure, though I see it's unavoidable here.

Reply:
This is only optimization but I think a pretty important one.

Reply:
I have added a paragraph in `doc.go`.